src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala

```diff
@@ -0,0 +1,243 @@
+            assert(
+              Try(children.last.asInstanceOf[ShuffleExchangeExec]).isSuccess || Try(
+                children.last.asInstanceOf[ReusedExchangeExec]).isSuccess)
+            assert(bucketSpec.numBuckets === 200)
+            p
+          case p @ FileSourceScanExec(_, _, _, _, _, dataFilters, _) =>
+            // Check filter pushed down properly.
+            assert(
+              dataFilters.toString.contains(" >= 2000)") && dataFilters.toString.contains(
+                " <= 4000)"))
+            p
+        }
+        assert(execNodes.count(p => p.isInstanceOf[BucketUnionExec]) === 2)
+        // 2 of index, 1 of appended file (1 is reused)
+        assert(execNodes.count(p => p.isInstanceOf[FileSourceScanExec]) === 3)
+
+        val join2 = query.join(query2, "clicks")
+        checkAnswer(join2, join)
+      }
+    }
+  }
+
+  test(
+    "Append-only: filter rule & json format, " +
+      "appended data should be shuffled and merged by Union") {
+    df = spark.read.json(sampleJsonDataLocation)
+    val query = df.filter(df("clicks") <= 2000).select(df("query"))
+
+    withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "false") {
+      val transformed = FilterIndexRule(query.queryExecution.optimizedPlan)
+      assert(transformed.equals(query.queryExecution.optimizedPlan), "No plan transformation.")
+    }
+
+    withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "true") {
+      val planWithHybridScan = FilterIndexRule(query.queryExecution.optimizedPlan)
+      assert(
+        !planWithHybridScan.equals(query.queryExecution.optimizedPlan),
+        "Plan should be transformed.")
+
+      // Check appended file is added to relation node or not.
+      val nodes = planWithHybridScan collect {
+        case p @ LogicalRelation(fsRelation: HadoopFsRelation, _, _, _) =>
+          val appendedFileCnt = fsRelation.location.inputFiles.count(_.contains(".copy"))
+          val indexFileCnt = fsRelation.location.inputFiles.count(_.contains("index2"))
+          assert(appendedFileCnt === 1 || indexFileCnt === 4)
+          assert(appendedFileCnt * indexFileCnt === 0)
+          p
+        case p @ Union(_) => p
+      }
+      assert(nodes.count(p => p.isInstanceOf[LogicalRelation]) === 2)
+      assert(nodes.count(p => p.isInstanceOf[Union]) === 1)
+      // Make sure there is no shuffle.
+      planWithHybridScan.foreach(f => assert(Try(f.asInstanceOf[Exchange]).isFailure))
+
+      spark.enableHyperspace()
+      val execPlan = spark.sessionState.executePlan(planWithHybridScan).executedPlan
+
+      val execNodes = execPlan collect {
+        case p @ UnionExec(children) =>
+          assert(children.size === 2)
+          // head is always the index plan.
+          assert(Try(children.head.asInstanceOf[WholeStageCodegenExec]).isSuccess)
+          assert(Try(children.last.asInstanceOf[WholeStageCodegenExec]).isSuccess)
```

Comment:
I think you need to update the comment to differentiate the behavior from the previous test.

Btw, you can turn off whole stage code gen if it makes the validation easier (e.g, `BucketedReadSuite.scala` does that)